### PR TITLE
fix(policy): improve targetRef name and tags validation

### PR DESCRIPTION
**Pull request by alparslanavci**
Added validation for targetRef name and tags as required. Also, MeshGateway and MeshHTTPRoute kind validations are included.

Fix https://github.com/kumahq/kuma/issues/7912

### Checklist prior to review

- [x] [Link to relevant issue][1] as well as docs and UI issues --
- [x] This will not break child repos: it doesn't hardcode values (.e.g "kumahq" as a image registry) and it will work on Windows, system specific functions like `syscall.Mkfifo` have equivalent implementation on the other OS --
- [x] Tests (Unit test, E2E tests, manual test on universal and k8s) --
- [x] Do you need to update [`UPGRADE.md`](../blob/master/UPGRADE.md)? --
- [x] Does it need to be backported according to the [backporting policy](../blob/master/CONTRIBUTING.md#backporting)? ([this](https://github.com/kumahq/kuma/actions/workflows/auto-backport.yaml) GH action will add "backport" label based on these [file globs](https://github.com/kumahq/kuma/blob/master/.github/workflows/auto-backport.yaml#L6), if you want to prevent it from adding the "backport" label use [no-backport-autolabel](https://github.com/kumahq/kuma/blob/master/.github/workflows/auto-backport.yaml#L8) label) --
- [x] Do you need to explicitly set a [`> Changelog:` entry here](../blob/master/CONTRIBUTING.md#submitting-a-patch) or add a `ci/` label to run fewer/more tests?

[1]: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

### CODE REVIEW

**Comment by slonka** [comment]
@bartsmykla @Automaat can you take a look at this? also @lobkovilya might be interested

File: pkg/plugins/policies/core/matchers/validators/validator.go [hunk 1]
```diff
@@ -34,18 +41,62 @@ func ValidateTargetRef(
...
 			verr.Add(requiredField("name", ref.Name, refKind))
+			verr.Add(validName(ref.Name))
 			verr.Add(disallowedField("mesh", ref.Mesh, refKind))
+			verr.Add(validTags(ref.Tags))
 		}
 	}
 	return verr
 }
 
+func validTags(tags map[string]string) validators.ValidationError {
```
Comment (bartsmykla) [inline pkg/plugins/policies/core/matchers/validators/validator.go:60]: we have a convention to name these kind of functions with `validate` prefix
```suggestion
func validateTags(tags map[string]string) validators.ValidationError {
```

File: pkg/plugins/policies/core/matchers/validators/validator.go [hunk 2]
```diff
@@ -34,18 +41,62 @@ func ValidateTargetRef(
...
+			res.Violations = append(res.Violations, validators.Violation{
+				Field:   "tags",
+				Message: `tag value must consist of alphanumeric characters, dots, dashes, slashes and underscores or be "*"`,
+			})
+		}
+	}
+	return res
+}
+
+func validName(value string) validators.ValidationError {
```
Comment (bartsmykla) [inline pkg/plugins/policies/core/matchers/validators/validator.go:89]:
```suggestion
func validateName(value string) validators.ValidationError {
```

File: pkg/plugins/policies/core/matchers/validators/validator.go [hunk 3]
```diff
@@ -2,11 +2,18 @@ package validators
...
+	"regexp"
 
 	common_api "github.com/kumahq/kuma/api/common/v1alpha1"
 	"github.com/kumahq/kuma/pkg/core/validators"
 )
 
+var (
+	nameCharacterSet     = regexp.MustCompile("^[0-9a-z-_]*$")
+	tagNameCharacterSet  = regexp.MustCompile(`^[a-zA-Z0-9\.\-_:/]*$`)
+	tagValueCharacterSet = regexp.MustCompile(`^([a-zA-Z0-9\.\-_:/]*|\*)$`)
```
Comment (bartsmykla) [inline pkg/plugins/policies/core/matchers/validators/validator.go:14]: I don't think you have to escape `.` and `-` inside `[]` when using backticks to describe regexp

**Review by bartsmykla — APPROVED** [review]
few nits but lgtm overall

File: pkg/plugins/policies/core/matchers/validators/validator.go (hunk 1 above)
Comment (alparslanavci) [reply pkg/plugins/policies/core/matchers/validators/validator.go]: Done 👍

File: pkg/plugins/policies/core/matchers/validators/validator.go (hunk 2 above)
Comment (alparslanavci) [reply pkg/plugins/policies/core/matchers/validators/validator.go]: Done 👍

File: pkg/plugins/policies/core/matchers/validators/validator.go (hunk 3 above)
Comment (alparslanavci) [reply pkg/plugins/policies/core/matchers/validators/validator.go]: I removed the escape for `.` but it is required for `-` to differentiate it from the range tick. Updated accordingly 👍

**Comment by alparslanavci** [comment]
Thanks for the review, @bartsmykla!

File: pkg/plugins/policies/core/matchers/validators/validator.go [hunk 4]
```diff
@@ -2,11 +2,18 @@ package validators
...
+	"regexp"
 
 	common_api "github.com/kumahq/kuma/api/common/v1alpha1"
 	"github.com/kumahq/kuma/pkg/core/validators"
 )
 
+var (
+	nameCharacterSet     = regexp.MustCompile(`^[0-9a-z-_]*$`)
+	tagNameCharacterSet  = regexp.MustCompile(`^[a-zA-Z0-9.\-_:/]*$`)
+	tagValueCharacterSet = regexp.MustCompile(`^([a-zA-Z0-9.\-_:/]*|\*)$`)
```
Comment (lahabana) [inline pkg/plugins/policies/core/matchers/validators/validator.go:14]: Should we make and use the ones:

https://github.com/kumahq/kuma/blob/573a9a5f8b569a6d5606bdab384156c34420e022/pkg/core/resources/apis/mesh/validators.go#L39-L41

More generally should we move this whole file in:

https://github.com/kumahq/kuma/blob/master/pkg/core/resources/apis/mesh/validators.go

or at least in `/resources/apis/` @bartsmykla what do you think?

File: pkg/plugins/policies/core/matchers/validators/validator.go [hunk 5]
```diff
@@ -34,18 +41,62 @@ func ValidateTargetRef(
...
 		}
 	}
 	return verr
 }
 
+func validateTags(tags map[string]string) validators.ValidationError {
+	res := validators.ValidationError{}
+	for key, value := range tags {
+		if key == "" {
+			res.Violations = append(res.Violations, validators.Violation{
```
Comment (lahabana) [inline pkg/plugins/policies/core/matchers/validators/validator.go:64]: Using `res.AddViolation` is a little easier. This enables you to add the key in the error:

```suggestion
			res.AddViolation(validators.RootedAt("tags").Key(key), validators.MustNotBeEmpty))
```

Note we also use existing error messages defined in:

https://github.com/kumahq/kuma/blob/573a9a5f8b569a6d5606bdab384156c34420e022/pkg/core/validators/messages.go#L8-L21

File: pkg/plugins/policies/core/matchers/validators/validator.go [hunk 6]
```diff
@@ -34,18 +41,62 @@ func ValidateTargetRef(
...
 			verr.Add(requiredField("name", ref.Name, refKind))
+			verr.Add(validateName(ref.Name))
 			verr.Add(disallowedField("mesh", ref.Mesh, refKind))
+			verr.Add(validateTags(ref.Tags))
 		}
 	}
 	return verr
 }
 
+func validateTags(tags map[string]string) validators.ValidationError {
```
Comment (lahabana) [inline pkg/plugins/policies/core/matchers/validators/validator.go:60]: Should this just be `validators.ValidateTags` ?

**Review by lahabana — COMMENTED** [review]
I think we're reintroducing things that already exists.

File: pkg/plugins/policies/core/matchers/validators/validator.go (hunk 4 above)
Comment (bartsmykla) [reply pkg/plugins/policies/core/matchers/validators/validator.go:14]: I like the idea to move it to https://github.com/kumahq/kuma/blob/master/pkg/core/resources/apis/mesh/validators.go

File: pkg/plugins/policies/core/matchers/validators/validator.go (hunk 6 above)
Comment (bartsmykla) [reply pkg/plugins/policies/core/matchers/validators/validator.go]: addressed

File: pkg/plugins/policies/core/matchers/validators/validator.go (hunk 5 above)
Comment (bartsmykla) [reply pkg/plugins/policies/core/matchers/validators/validator.go]: addressed

**Comment by bartsmykla** [comment]
@alparslanavci I allowed myself to address @lahabana remarks. Please have a look if everything is ok for you and I would merge it then.

**Comment by alparslanavci** [comment]
Thanks for the help, @bartsmykla! I was a bit busy in the last few days, couldn't find a chance to address the comments. I am ok to go ahead 👍 

(I think we tried to fix the failing tests at the same time, thus I reverted my latest commit.)

**Pull request merged**